### ISSUE TITLE
Zero or more path query is broken with some datasets

### ISSUE THREAD

**Issue by afirmage** [eclipse/rdf4j#689]
We've just upgraded our software from using Sesame 2.7.15 to rdf4j 2.1.2 and found that property paths (https://www.w3.org/TR/sparql11-property-paths/#path-language) that use * do not work with some datasets. I had trouble building up a dataset from scratch to reproduce the issue (it seemed to work on smaller datasets) so instead took the larger dataset that we did have that was showing the issue and have attached that (after scrambling some IP out of it). There are screen captures showing the steps to reproduce the issue and the dataset and queries.

We have been able to work around the issue by re-writing the query using a UNION clause, but have now found another query that doesn't seem to work, even after doing that. At this point we can't upgrade as we are not confident that the queries will continue to work correctly as the underlying dataset changes.

[rdf4j-bug.zip](https://github.com/eclipse/rdf4j/files/635373/rdf4j-bug.zip)

**Comment by ansell** [comment]
Hmmm, the tests Jeen added when I was having issues with property paths myself years ago only include +, not *:

https://github.com/eclipse/rdf4j/blob/master/testsuites/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/ComplexSPARQLQueryTest.java#L1469

**Comment by jeenbroekstra** [comment]
Oddly, I can reproduce the problem when executing the query from the RDF4J workbench, but not when executing directly from code (on a SailRepository). So it appears the SPARQL engine itself works as expected, but something is going wrong in communicating the query _to_ the engine.

**Comment by jeenbroekstra** [comment]
Disregard my previous comment, made a mistake in trying to duplicate the issue. I can reproduce from the workbench as well as directly from code.

**Comment by jeenbroekstra** [comment]
It appears to be related to how the query planner optimizes the joins. This is why the problem is hard to reproduce with a smaller dataset.

**Comment by afirmage** [comment]
Fast work guys! I'll pick up 2.1.4 when released and retest. Is it possible to download the current build before then to test it?

**Comment by jeenbroekstra** [comment]
You can grab a [nightly snapshot build](http://rdf4j.org/download/#Source_code_and_nightly_builds), though keep in mind that that is from our development branch, so it may also include some changes that will not be part of the next patch release.

**Comment by afirmage** [comment]
Hi Jeen,
We've tested the updated jar file and it seems to have fixed the issue! Nice! I noticed that this issue was scheduled for milestone build 2.1.4. Is this going to be an official released package as we were hoping to pick it up when it was closed off, or do we need to wait for 2.2.0 in late January?

**Comment by jeenbroekstra** [comment]
I'm way ahead of you. Patch release 2.1.4 is now available for download, this fix is part of it. I'm writing the release announcement e-mail as we speak :)

**Comment by afirmage** [comment]
Nice work! Thanks for getting a package out so quickly!

**Comment by geekysquirrel** [comment]
I know this has been closed but I'm still experiencing tha same issue. It's been driving me mad for some time now. I have a more complex model but I narrowed it down to this.
During execution of my Java code, I print the  query which returns no results when executed via RDF4J. However, if I execute the same query using a third party tool such as Topbraid Composer, I get the desired results. I upgraded to 2.1.4 but it didn't seem to have made a difference for me.
My dataset is not very large, a couple of thousand triples only. It happens when checking for rdfs:subClassOf*.
Any ideas?
